### PR TITLE
[Wav2Vec2] Fix special tokens for Wav2Vec2 tokenizer

### LINKED ISSUE #10942: Wav2Vec2CTCTokenizer does not take the vocabulary into account when identifying tokens in a sentence
## Environment info
- `transformers` version: 4.4.0
- Platform: Darwin-19.6.0-x86_64-i386-64bit
- Python version: 3.7.9
- PyTorch version (GPU?): 1.8.1 (False)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help
@patrickvonplaten @LysandreJik

## Information

Model I am using (Bert, XLNet ...): wav2vec2

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [X] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [X] my own task or dataset: (give details below)

## To reproduce

We are trying to train an automatic phonemic transcription system for a low-resource language using the instructions [here](https://huggingface.co/blog/fine-tune-xlsr-wav2vec2 )

I created a `Wav2Vec2CTCTokenizer` tokenizer as follows:
```python
from transformers import Wav2Vec2CTCTokenizer

# example of a phonemic transcription                                                                         
sent = "ʈʰ æ æ̃ ˧ kʰ"
# phonemes are separated by spaces in the transcription                                                       
vocab = {phoneme for phoneme in sent.split()}
vocab_dict = {k: v for v, k in enumerate(vocab)}

# <space> will be our phoneme separator                                                                       
vocab_dict[" "] = len(vocab_dict)
vocab_dict["[UNK]"] = len(vocab_dict)
vocab_dict["[PAD]"] = len(vocab_dict)

import json
with open('vocab.json', 'w') as vocab_file:
    json.dump(vocab_dict, vocab_file)

tokenizer = Wav2Vec2CTCTokenizer("./vocab.json",
                                 unk_token="[UNK]",
                                 pad_token="[PAD]",
                                 word_delimiter_token=" ")

```

The result of the sentence tokenization is:
```
>>> tokenizer(sent)

{'input_ids': [6, 6, 5, 2, 5, 2, 6, 5, 4, 5, 6, 6], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

```
Where 6 is the id of the <unk> token. The vocabulary is :
`{'æ̃': 0, 'kʰ': 1, 'æ': 2, 'ʈʰ': 3, '˧': 4, ' ': 5, '[UNK]': 6, '[PAD]': 7}`

It appears that phonemes made of several characters (e.g. ʈʰ) are not recognized as a whole but rather taken separately (ʈ and than ʰ) each token being mapped to a separate id (here `<unk>` as the separated characters are not in the vocabulary). 

The tokenization output results from the `Wav2Vec2CTCTokenizer._tokenize` function being called before looking in the dictionary representing the vocabulary to map tokens into IDs. This function converts the string representing the sentence to tokenize into a list without taking into account the tokens defined in the vocabulary.

I do not know if this is the intended behaviour or if we are not using the tokenizer correctly (in which case the documentation might be improved)

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR fixes https://github.com/huggingface/transformers/issues/10942. Wav2Vec2's vocabulary can 
consist of multi-character tokens which should then nevertheless be treated as single atomic tokens when encoding/decoding. 
=> This PR ensures such behavior and fixes the issue attached to this PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by patil-suraj — APPROVED** [review]
Nice, LGTM!

File: src/transformers/models/wav2vec2/tokenization_wav2vec2.py [hunk 1]
```diff
@@ -142,6 +142,12 @@ def __init__(
             self.encoder = json.load(vocab_handle)
         self.decoder = {v: k for k, v in self.encoder.items()}
 
+        # make sure that tokens made of several
+        # characters are not split at tokenization
+        for token in self.encoder.keys():
+            if len(token) > 1:
+                self.unique_no_split_tokens.append(token)
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/tokenization_wav2vec2.py:149]: I don't understand this part: `unique_no_split` is intended for use with special tokens only, yet you put every tokens in it here.

Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/tokenization_wav2vec2.py:149]: `Wav2Vec2CTCTokenizer` is supposed to work have vocabularies that consist of characters and/or phonemes, but not subwords. Hence, I assumed that all tokens of the vocabulary consists of only one unicode character. However, there are some exceptions, such as: `"æ̃"` => this phoneme consists of two unicode characters and thus leads to an error as described in the issue. To solve the problem, I put all characters of length > 1 in `self.unque_no_split_tokens`... => is there a better way to handle it here (cc @n1t0)?

Comment (sgugger) [reply src/transformers/models/wav2vec2/tokenization_wav2vec2.py:149]: Thanks for the explanation, it makes sense to me!

**Review by LysandreJik — APPROVED** [review]

**Pull request merged**